_posts/2023-07-05-niklas-fuzzing.md

```diff
@@ -0,0 +1,84 @@
+---
+title: "Niklas Gögge on Fuzz Testing"
+permalink: /blog/2023/07/05/fuzzing/
+layout: post
+author: brink
+name: "brink"
+alt: Niklas Gögge on Fuzz Testing
+category: "News"
+description: Niklas Gögge on Fuzz Testing.
+---
+
+In a security-focused project like Bitcoin Core, testing is critical. [Bitcoin
+Core unit tests][] are written in C++ using the Boost libraries to quickly test
+components of the codebase in isolation (see the [bech32 unit tests][] for
+example). [Bitcoin Core functional tests][] are written in python to test
+features of the software, involving multiple components working together (see
+the [P2P ping functional tests][]).
+
+A more cutting-edge type of testing is fuzz testing where software repeatedly calls
+into portions of the Bitcoin Core codebase with random, potentially
+malformed or semi-valid inputs in order to surface potential negative consequences
+such as a crash, memory leak, or other failure scenarios. When done
+continuously, these auto-generated mutated inputs from the fuzzing engine can be
+used to expand the test coverage and uncover bugs.
+
+For example, fuzzing can uncover a series of specially-crafted P2P messages that
+leads to a crash, memory leak, or inconsistent internal state. As Bitcoin Core
+nodes are constantly exposed to anonymous peers on the network, a crash bug in
+the P2P message processing code could be catastrophic.
+
+In contrast to manual testing or individual code review, fuzz testing scales
+with the amount of computation allocated to it and can be running 24/7. Bitcoin
+Core participates in [OSS-Fuzz][], Google's continuous fuzzing program, in
+addition to contributors running fuzzing infrastructure and contributing new
+inputs back to the growing body of [fuzzing corpora][].
```

Comment:
```suggestion
As a form of automated software testing, fuzzing can scale (to some extend) with the amount of computation allocated to it and if done continuously often finds edge cases that were missed during manual code review. Bitcoin
Core participates in [OSS-Fuzz][], Google's continuous fuzzing program, in
addition to contributors running fuzzing infrastructure and contributing new
inputs back to the growing body of [fuzzing corpora][].
```

Reply:
Thanks @dergoegge , took this suggestion with a small change